src/Obelisk/Frontend/GoogleAnalytics.hs

```diff
@@ -1,54 +1,250 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
 {-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -Wall #-} -- -Werror #-}
 
 module Obelisk.Frontend.GoogleAnalytics
   ( googleAnalyticsFromConfig
   , googleAnalytics
   , trackingIdPath
   , getTrackingId
+  , Analytics(..)
+  , GoogleAnalyticsT(..)
+  , GtagJSCall(..)
+  , gaClickEvent
+  , gaOutboundClickEvent
+  , runGoogleAnalyticsT
   ) where
 
-import Control.Monad (void)
+import Control.Monad
+import Control.Monad.Exception
+import Control.Monad.Fix
+import Control.Monad.IO.Class
+import Control.Monad.Primitive
+import Control.Monad.Ref
+import Control.Monad.State.Strict
 import Data.Text (Text)
-import GHCJS.DOM.Types (liftJSM)
 import Language.Javascript.JSaddle.Evaluate (eval)
 import Obelisk.Configs
 import Obelisk.Route.Frontend
 import Reflex.Dom.Core
 import qualified Data.Text as T
+import qualified Data.Aeson as Aeson
+import Data.Sequence (Seq)
+import qualified Data.Sequence as Seq
+import Language.Javascript.JSaddle
+import Reflex.Host.Class
+
+import qualified Data.HashMap.Strict as HMap
+
+class Monad m => Analytics t w m | m -> w where
+  tellAnalytics :: Event t w -> m ()
+  -- ^ Tell analytics something, with asynchronous delivery.
+
+  -- tellAnalyticsSync :: Event t w -> m ()
+  -- ^ Tell analytics something, with synchronous delivery.
+  --   This waits until the message is delivered before returning.
+
+  -- tellAnalyticsWithSync :: Event t w -> m (STM Bool)    <- this type is conceptually flawed
+  -- ^ Tell analytics something, with asynchronous delivery.
+  --   Returns a 'STM' action that becomes 'True' after delivery of this particular message.
+  --   This is the preferred way to implement synchronous delivery with timeout.
+
+  -- syncAnalytics :: m ()
+  -- ^ Synchronize on the delivery of every message outstanding at the time of the call.
+  --   This returns after all such messages have been delivered.
+
+  -- asyncAnalytics :: m (STM Bool)
+  -- ^ Returns an 'STM' action that becomes 'True' after all currently outstanding messages have been delivered.
+  --   This is an asynchronous variant of 'syncAnalytics', and the preferred way of doing a "global" sync with timeout.
```

Comment:
All references to `STM Bool` are not FRP-y. They should be something like `Dynamic Bool`

Reply:
How do you implement sync with sane timeouts without STM, then?  My idea is that you create a `TVar` for each message, that `gtag.js` sets to true by calling back into the Haskell runtime.   Then the action that is somehow the result of that message is simply,  read that` TVar`.   (We shouldn't let people write that `TVar`.)

However, that's more an issue with the conceptually flawed `tellAnalyticsWithSync`:  a "global" this avoids some of that complication.   (Though, a global sync shouldn't wait for the completion of messages that are outstanding at the time of completion but weren't outstanding at the time it was called.)  

None of the current external links navigate away from the page, so I'm not aware of any immediate need for implementing synchronous delivery of any sort.

That's more an issue with